pkg/apis/networking/types.go

```diff
@@ -544,3 +544,68 @@ type ServiceBackendPort struct {
 	// +optional
 	Number int32
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ClusterCIDRConfig is the Schema for the clustercidrconfigs API.
+type ClusterCIDRConfig struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec   ClusterCIDRConfigSpec
+	Status ClusterCIDRConfigStatus
+}
+
+// ClusterCIDRConfigSpec defines the desired state of ClusterCIDRConfig.
+type ClusterCIDRConfigSpec struct {
+	// NodeSelector defines which nodes the config is applicable to.
+	// An empty or nil NodeSelector functions as a default that applies to all nodes.
+	// This field is immutable.
+	// +optional
+	NodeSelector *api.NodeSelector
```

Comment:
I have few questions with regards to this field:
1. What happens if two `ClusterCIDRConfig`s of the same IPFamily apply on the same node or set of nodes?
2. What happens if a `ClusterCIDRConfig` instance apply on node X then user modifies `.Spec.NodeSelector` where node X is no longer part of the selector (or changing node labels).

Reply:
I think #2 here could be really tricky. If the config no longer applies to a node, do we drain pods using that cidr gracefully? If so, we'd need to actually use Status to show if the config has been fully reconciled with all nodes' state. And block assigning the old node's range to a new node until the old node got drained.

Or something like that?

Reply:
regarding #1, we've enumerated pretty extensive tie-break rules: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2593-multiple-cluster-cidrs#expected-behavior  (see bullets 4 and 5)
So at the very least the behavior should be deterministic.

regarding #2, the controller won't (and shouldn't) take away to PodCIDR after it's been written in Node.Spec. Plenty of CNIs and other components will misbehave if the value changes during the lifetime of the node.
The selector is supposed to be immutable, however, the labels on the node can be changed as you pointed out. The controller doesn't take away the PodCIDR after it's been allocated -- we sort of make reference to that in a couple places [1] & [2]. So if I edit the node labels so it's no longer selected, then the node keeps its old PodCIDR allocation. The controller won't allocate that range to any other node, as it might still be in use. If you as the user manually clear the PodCIDR allocation from the Node.Spec, then the controller considers that range free and allocatable to another node.

[1]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2593-multiple-cluster-cidrs#notesconstraintscaveats
[2]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2593-multiple-cluster-cidrs#event-watching-loops

Reply:
Can you describe the tie -breaking in the API doc? where will this documentation live long-term (KEP doesn't seem like the best place for this to be long-term).

Reply:
Added the tie breaking rules to the API doc